File: utils/mpmt/mpmt1.pl
#!/usr/bin/env perl
#
# mpmt1.pl: Perl version of mpmt1.py
#
# License:
#   Apache License, Version 2.0
# History:
#   * 2022/04/16 v0.1 Initial version
# Author:
#   Masanori Itoh <masanori.itoh@gmail.com>
# TOTO:
#   * Implement multi process model
use strict;
use Getopt::Std;
use Time::HiRes;
use threads;


my $busy_worker = sub
{
    my $idx = shift(@_);
    my $duration = shift(@_);
    printf "busy_worker: %d\n", $idx;
    my $ts_start;
    my $ts_now;
    my $continue = 1;
    $ts_start = Time::HiRes::gettimeofday;
    while ($continue) {
	$ts_now = Time::HiRes::gettimeofday;
	if (($ts_now - $ts_start) > $duration) {
	    printf "busy_worker: exiting... %s\n", $ts_now - $ts_start;
	    $continue = 0;
	}
    }
};
#
# main routine
#
my @th;
my $num_context = 2;
my $duration = 5.0;
my $mode = 'thread';
my %opt;

getopts('n:d:m:', \%opt);
if (exists $opt{n}) {
    $num_context = $opt{n};
}
if (exists $opt{d}) {
    $duration = $opt{d};
}
printf "mpmt1.pl: num_context = %d duration = %f\n", $num_context, $duration;

for (my $i = 0; $i < $num_context; $i++) {
    $th[$i] = threads->new(\&$busy_worker, $i, $duration)
}

for (my $i = 0; $i < $num_context; $i++) {
    $th[$i]->join;
}

File: utils/mpmt/mpmt1.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-
#
# mpmt1.rb: A Ruby version of mpmt1.py
#
# License:
#   Apache License, Version 2.0
# History:
#   * 2021/12/29 v0.1 Initial version
# Author:
#   Masanori Itoh <masanori.itoh@gmail.com>
# TODO:
#   * Implement multi process mode (or more clever thread mode)


def busy_worker(id, num_context, duration)

  printf("thread: %d / n: %d d: %d %s\n", id, num_context, duration, Thread.current)

  tsnow = Time.now
  ts_save = tsnow.to_i * 1000*1000 + tsnow.usec

  loop do
    tsnow = Time.now
    ts = tsnow.to_i * 1000*1000 + tsnow.usec
    if (ts - ts_save) > duration
      printf("thread: %d Expired! %d / %d\n",  id, (ts- ts_save), duration)
      return
    end
  end

end


if $0 == __FILE__

  require 'optparse'

  num_context = 4
  duration = 10 * 1000 * 1000
  mode = 't'

  params = ARGV.getopts("n:", "d:", "m:")

  if params.has_key? 'n'
    num_context = params['n'].to_i
  end
  if params.has_key? 'd'
    duration = params['d'].to_i
    duration = duration * 1000 * 1000
  end
  if params.has_key? 'm'
    mode = params['m']
  end

  printf("num_context: %d duration: %d (us) mode: %s (t: thread)\n", num_context, duration, mode)

  if mode == 't' or mode == 'T'
    threads = []

    num_context.times do |idx|
      threads << Thread.new { busy_worker(idx, num_context, duration) }
    end

    threads.each() { |th|
      p th.join()
    }

  elsif mode == 'p' or mode == 'P'
    printf("multi process mode is not yet implemented: %s\n", mode)
    exit
  else
    printf("Invalide mode: %s\n", mode)
    exit
  end

end
